File: Algorithms/Week1_Coursera/UnionFind/ResizingArrayStackOfStrings.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace UnionFind
{
    public class ResizingArrayStackOfStrings
    {
        string[] s;
        int N = 1;
        public ResizingArrayStackOfStrings()
        {
            s = new String[N];
        }

        public void push(String item)
        {
            if (N == s.Length) 
                resize(2 * s.Length);
            s[N++] = item;
        }

        public String pop()
        {
            String item = s[--N];
            s[N] = null;
            if (N > 0 && N == s.Length / 4) resize(s.Length / 2);
            return item;
        }


        private void resize(int capacity)
        {
            String[] copy = new String[capacity];
            for (int i = 0; i < N; i++)
                copy[i] = s[i];
            s = copy;
        }

    }
}

File: CourseraJava/Week2_Stack_Queue/src/ResizingArrayStackOfStrings.java
/**
 * Created by ivko on 3/11/14.
 */
public class ResizingArrayStackOfStrings {

    String[] s;
    int N =0;

    public ResizingArrayStackOfStrings(){
        s = new String[1];
    }

    public void push(String item){
        if (N == s.length)
            resize(2 * s.length);
        s[N++] = item;
    }

    public String pop()
    {
        String item = s[--N];
        s[N] = null;
        if (N > 0 && N == s.length/4) resize(s.length/2);
        return item;
    }

    private void resize(int capacity){
        String[] copy = new String[capacity];
        for (int i = 0; i < N; i++)
            copy[i] = s[i];
        s = copy;
    }

    public  static  void  main(String[] args){
        ResizingArrayStackOfStrings stack = new ResizingArrayStackOfStrings();

        String inputString = "to be or not to - be - - that - - - is";

        String[] strArr = inputString.split(" ");

        for(String str : strArr){
            if (str.equals("-"))
                System.out.print(" " + stack.pop());
            else
                stack.push(str);

    }
    }
}
